video_xblock/static/js/student-view/player-state.js

```diff
@@ -0,0 +1,130 @@
+                url: transcript.url
+            };
+        });
+        return result;
+    };
+
+    var transcripts = getTranscipts(playerStateObj.transcripts);
+
+    /** Restore default or previously saved player state */
+    var setInitialState = function(state) {
+        var stateCurrentTime = state.currentTime;
+        var playbackProgress = localStorage.getItem('playbackProgress');
+        if (playbackProgress) {
+            playbackProgress = JSON.parse(playbackProgress);
+            if (playbackProgress[window.videoPlayerId]) {
+                stateCurrentTime = playbackProgress[window.videoPlayerId];
+            }
+        }
+        if (stateCurrentTime > 0) {
+            player.currentTime(stateCurrentTime);
+        }
+        player
+            .volume(state.volume)
+            .muted(state.muted)
+            .playbackRate(state.playbackRate);
+        player.transcriptsEnabled = state.transcriptsEnabled;  // eslint-disable-line no-param-reassign
+        player.captionsEnabled = state.captionsEnabled;  // eslint-disable-line no-param-reassign
+        player.captionsLanguage = state.captionsLanguage;  // eslint-disable-line no-param-reassign
+        // To switch off transcripts and captions state if doesn`t have transcripts with current captions language
+        if (!transcripts[player.captionsLanguage]) {
+            player.captionsEnabled = player.transcriptsEnabled = false; // eslint-disable-line no-param-reassign
+        }
+    };
+
+    /**
+     * Save player state by posting it in a message to parent frame.
+     * Parent frame passes it to a server by calling VideoXBlock.save_state() handler.
+     */
+    var saveState = function() {
+        var playerObj = this;
+        var transcriptUrl = getDownloadTranscriptUrl(transcripts, playerObj);
+
+        var newState = {
+            volume: playerObj.volume(),
+            currentTime: playerObj.ended() ? 0 : playerObj.currentTime(),
+            playbackRate: playerObj.playbackRate(),
+            muted: playerObj.muted(),
+            transcriptsEnabled: playerObj.transcriptsEnabled,
+            captionsEnabled: playerObj.captionsEnabled,
+            captionsLanguage: playerObj.captionsLanguage
+        };
+        if (JSON.stringify(newState) !== JSON.stringify(playerState)) {
+            console.log('Starting saving player state');  // eslint-disable-line no-console
+            playerState = newState;
+            parent.postMessage({
+                action: 'saveState',
+                info: newState,
+                xblockUsageId: xblockUsageId,
+                downloadTranscriptUrl: transcriptUrl || '#'
+            },
+            document.location.protocol + '//' + document.location.host
+            );
+        }
+    };
+
+    /**
+     *  Save player progress in browser's local storage.
+     *  We need it when user is switching between tabs.
+     */
+    var saveProgressToLocalStore = function saveProgressToLocalStore() {
```

Comment:
hm, I didn't know about debugging. If it's really used during debugging, it's ok.